src/cockpit/samples.py

```diff
@@ -241,6 +240,30 @@ def sample(self, samples):
                 usage_usec = self.read_cgroup_integer_stat(rootfd, 'cpu.stat', True, key=b'usage_usec')
                 if usage_usec:
                     samples['cgroup.cpu.usage'][cgroup] = usage_usec / 1000
+        else:
+            memory_path = '/sys/fs/cgroup/memory/'
+            for path, _, _, rootfd in os.fwalk(memory_path):
+                cgroup = path.replace(memory_path, '')
+
+                if not cgroup:
+                    continue
+
+                samples['cgroup.memory.usage'][cgroup] = self.read_cgroup_integer_stat(rootfd, 'memory.usage_in_bytes', True)
+                samples['cgroup.memory.limit'][cgroup] = self.read_cgroup_integer_stat(rootfd, 'memory.limit_in_bytes')
+                samples['cgroup.memory.sw-usage'][cgroup] = self.read_cgroup_integer_stat(rootfd, 'memory.memsw.usage_in_bytes', True)
+                samples['cgroup.memory.sw-limit'][cgroup] = self.read_cgroup_integer_stat(rootfd, 'memory.memsw.limit_in_bytes')
+
+            cpu_path = '/sys/fs/cgroup/cpu/'
+            for path, _, _, rootfd in os.fwalk(cpu_path):
```

Comment:
Fun fact: `fwalk()` itself has a `dir_fd` parameter.

Less fun fact: the pathnames that it returns are prefixed by `./`, so you'd need to trim them anyway :(

Reply:
Interesting, that would also mean wrapping it in a `try`/`finally` to handle the `os.close`.